src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalPane.java

```diff
@@ -112,42 +117,179 @@ public void createTerminal()
          secureInput_ = new ShellSecureInput();  
       }
       
-      TerminalSession newSession = new TerminalSession(secureInput_);
+      TerminalSession newSession = new TerminalSession(secureInput_,
+                                                       nextTerminalSequence());
       newSession.connect();
    }
 
    @Override
    public void onTerminalSessionStarted(TerminalSessionStartedEvent event)
    {
-      terminalSessionsPanel_.add(event.getTerminalWidget());
-      terminalSessionsPanel_.showWidget(event.getTerminalWidget());
+      TerminalSession terminal = event.getTerminalWidget();
       
-      // TODO (gary) update dropdown
-   } 
+      terminalSessionsPanel_.add(terminal);
+      terminalSessionsPanel_.showWidget(terminal);
+      setFocusOnVisible();
+   }
    
    @Override
    public void onTerminalSessionStopped(TerminalSessionStoppedEvent event)
    {
-      Widget currentTerminal = event.getTerminalWidget();
+      TerminalSession currentTerminal = event.getTerminalWidget();
       int currentIndex = terminalSessionsPanel_.getWidgetIndex(currentTerminal);
-      if (currentIndex > 0)
+      int newIndex = terminalToShowWhenClosing(currentIndex);
+      if (newIndex >= 0)
       {
-         terminalSessionsPanel_.showWidget(currentIndex - 1);
-         // TODO (gary) set focus on the widget
+         terminalSessionsPanel_.showWidget(newIndex);
+         setFocusOnVisible();
       }
       terminalSessionsPanel_.remove(currentTerminal);
-      
-      // TODO (gary) update dropdown
+
+      if (terminalSessionsPanel_.getWidgetCount() < 1)
+      {
+         activeTerminalToolbarButton_.setNoActiveTerminal();
+         setTerminalCaption("");
+      }
    }
 
+   @Override
+   public void onSwitchToTerminal(SwitchToTerminalEvent event)
+   {
+      TerminalSession terminal = terminalWithHandle(event.getTerminalHandle());
+      if (terminal != null)
+      {
+         terminalSessionsPanel_.showWidget(terminal);
+         setFocusOnVisible();
+      }
+   }
+   
+   @Override
+   public void onTerminalCaption(TerminalCaptionEvent event)
+   {
+      TerminalSession visibleTerm = getVisibleTerminal();
+      TerminalSession captionTerm = event.getTerminalSession();
+      if (visibleTerm != null && visibleTerm.getHandle().equals(
+            captionTerm.getHandle()))
+      {
+         setTerminalCaption(captionTerm.getCaption());
+      }
+   }
+   
+   /**
+    * @return number of terminals hosted by the pane
+    */
    public int getTerminalCount()
    {
       return terminalSessionsPanel_.getWidgetCount();
    }
+   
+   /**
+    * @param i index of terminal to return
+    * @return terminal at index, or null
+    */
+   public TerminalSession getTerminalAtIndex(int i)
+   {
+      Widget widget = terminalSessionsPanel_.getWidget(i);
+      if (widget != null)
+      {
+         return (TerminalSession)widget;
```

Comment:
for resiliency, consider checking with `instanceof` that you're actually dealing with a `TerminalSession` here (otherwise this will throw)